### PR TITLE
feat(entire-app): This change allows the user to select the headers that he does not want Keploy to record

### PR DESCRIPTION
**Pull request by PranshuSrivastava**
Signed-off-by: Pranshu Srivastava <iampranshu24@gmail.com>

### CODE REVIEW

**Comment by slayerjain** [comment]
@PranshuSrivastava does this only work for headers or also for json keys in the body? Also how to differentiate between request vs response keys?

**Comment by PranshuSrivastava** [comment]
> @PranshuSrivastava does this only work for headers or also for json keys in the body?
> 
> Also should we differentiate between request vs response keys?

@slayerjain No, currently we are just filtering the headers using this list and this only works on the req property and not the res.

**Comment by slayerjain** [comment]
@thunderboltsid do you think it would be useful to do this for response fields and also body params if they’re json?

**Comment by thunderboltsid** [comment]
The fields we tend to twiddle almost always are:
- authorization headers (req & metadata): we remove these altogether
- Session cookie headers (in response): we remove these altogether
- domain name (req & metadata): we twiddle this out to a dummy domain
So, the generated mock looks like this
```yaml
version: api.keploy.io/v1beta2
kind: Http
name: TestMetaOperations_GetVersion
spec:
    metadata:
        Header: map[Accept:[application/json] Authorization:[Basic dXNlcjpwYXNzCg==] Content-Type:[application/json] User-Agent:[nutanix]]
        Proto: HTTP/1.1
        ProtoMajor: "1"
        ProtoMinor: "1"
        URL: https://internal.domain.tld:9440/endpoint
        name: http-client
        operation: GET
        type: HTTP_CLIENT
    req:
        method: GET
        proto_major: 1
        proto_minor: 1
        url: https://internal.domain.tld:9440/endpoint
        header:
            Accept: application/json
            Authorization: Basic dXNlcjpwYXNzCg==
            Content-Type: application/json
            User-Agent: nutanix
        body: ""
        body_type: utf-8
    resp:
        status_code: 200
        header:
            Content-Security-Policy: 'default-src ''self'' https://*.nutanix.com; script-src ''self'' ''unsafe-inline'' ''unsafe-eval''; style-src ''self'' ''unsafe-inline''; connect-src ''self'' wss: https://downloads.frame.nutanix.com https://downloads.frame.nutanix.us; img-src ''self'' blob: data:; frame-src ''self'' https://*.nutanix.com blob: data:'
            Content-Type: application/json
            Date: Tue, 24 Jan 2023 09:57:31 GMT
            Server: envoy
            Set-Cookie: NTNX_IGW_SESSION=2E5F0FE4B7FCAB78CD967B46AB32915CCBFDC85316FAFB9174BDC4A3745339D8AD9C1D70A8EDD3D4F06107F3B1A67E85AEB1CB04F563818BC5D029D9E956652ED0F8F93E7A7EFD1E2E59DF1D378424964541E2C9B8A41436D0092A4114FBC4FA702A94C5090C34D97A3310F88CD3E9BDC8553230343BB1E4641F848102DF9B96AA81D6407273F774CFBD3E3FD68CAF84654B21E177A36DE7FB09EFC5A50A5253E1B5F5D6C5082A01D756CC68680FE1FBCA64CD72F1D6B1865C6FFC4B6BB8DE2C2B510731103F18DDA2289515F604632A26BD68B0E03647B854243EAFD7264B0765F45279BD6735E26D58DF720D5B9061EF316730;Expires=Tue, 24-Jan-2023 10:12:30 GMT;Path=/;Secure;HttpOnly;SameSite=Lax
            Strict-Transport-Security: max-age=8640000:includeSubdomains
            Vary: Accept-Encoding
            X-Content-Type-Options: nosniff
            X-Dns-Prefetch-Control: "off"
            X-Envoy-Upstream-Service-Time: "665"
            X-Frame-Options: SAMEORIGIN
            X-Ntnx-Env: pc
            X-Ntnx-Product: pc.2022.6.0.1
            X-Permitted-Cross-Domain-Policies: master-only
            X-Powered-By: Express
            X-Xss-Protection: 1; mode=block
        body: |
            {"build_date":"2022-11-11T18:22:09+0000","version":"2.6.0"}
        body_type: utf-8
        status_message: ""
        proto_major: 0
        proto_minor: 0
    objects:
        - type: error
          data: H4sIAAAAAAAA/wEAAP//AAAAAAAAAAA=
```
and then we currently twiddle it to:
```yaml
version: api.keploy.io/v1beta2
kind: Http
name: TestMetaOperations_GetVersion
spec:
    metadata:
        Header: map[Accept:[application/json] Content-Type:[application/json] User-Agent:[nutanix]]
        Proto: HTTP/1.1
        ProtoMajor: "1"
        ProtoMinor: "1"
        URL: https://internal.domain.tld:9440/endpoint
        name: http-client
        operation: GET
        type: HTTP_CLIENT
    req:
        method: GET
        proto_major: 1
        proto_minor: 1
        url: https://internal.domain.tld:9440/endpoint
        header:
            Accept: application/json
            Content-Type: application/json
            User-Agent: nutanix
        body: ""
        body_type: utf-8
    resp:
        status_code: 200
        header:
            Content-Security-Policy: 'default-src ''self'' https://*.nutanix.com; script-src ''self'' ''unsafe-inline'' ''unsafe-eval''; style-src ''self'' ''unsafe-inline''; connect-src ''self'' wss: https://downloads.frame.nutanix.com https://downloads.frame.nutanix.us; img-src ''self'' blob: data:; frame-src ''self'' https://*.nutanix.com blob: data:'
            Content-Type: application/json
            Date: Tue, 24 Jan 2023 09:57:31 GMT
            Server: envoy
            Set-Cookie: Expires=Tue, 24-Jan-2023 10:12:30 GMT;Path=/;Secure;HttpOnly;SameSite=Lax
            Strict-Transport-Security: max-age=8640000:includeSubdomains
            Vary: Accept-Encoding
            X-Content-Type-Options: nosniff
            X-Dns-Prefetch-Control: "off"
            X-Envoy-Upstream-Service-Time: "665"
            X-Frame-Options: SAMEORIGIN
            X-Ntnx-Env: pc
            X-Ntnx-Product: pc.2022.6.0.1
            X-Permitted-Cross-Domain-Policies: master-only
            X-Powered-By: Express
            X-Xss-Protection: 1; mode=block
        body: |
            {"build_date":"2022-11-11T18:22:09+0000","version":"2.6.0"}
        body_type: utf-8
        status_message: ""
        proto_major: 0
        proto_minor: 0
    objects:
        - type: error
          data: H4sIAAAAAAAA/wEAAP//AAAAAAAAAAA=
```

**Comment by slayerjain** [comment]
Thanks for your response @thunderboltsid and tje examples :)

**Comment by PranshuSrivastava** [comment]
Hey @thunderboltsid thanks for the clear and concise explanation. The first feature is partially implemented as you can remove any headers from the req body. I will implement the other features and then you can take a look. 
Thanks

File: go.mod [hunk 1]
```diff
@@ -2,6 +2,8 @@ module github.com/keploy/go-sdk
 
 go 1.17
 
+replace go.keploy.io/server => ../keploy/
```
Comment (re-Tick) [inline go.mod:5]: Please remove or comment this statement.

Comment (PranshuSrivastava) [reply go.mod:5]: @re-Tick I was planning to remove the statement after the PR was ready to merge, as this allows the reviewers to test the changes without any problems. I have tried building this on my system, it builds without any errors.

**Comment by slayerjain** [comment]
The bulk of the code for this PR is in the keploy server repo - https://github.com/keploy/keploy/pull/289

**Pull request merged**